### ISSUE TITLE
PlugUpdate bugged on Windows

### ISSUE THREAD

**Issue by orlp** [junegunn/vim-plug#229]
Attempting to do `:PlugUpdate`, even with merely one plugin results in the following error on Windows:

    Error detected while processing function <SNR>3_update..<SNR>3_update_impl..<SNR>3_update_vim..<SNR>3_tick..<SNR>3_spawn..<SNR>3_system:
    line    7:
    E484: Can't open file C:\Users\Orson\AppData\Local\Temp\VIo5763.tmp
    Error detected while processing function <SNR>3_update..<SNR>3_update_impl..<SNR>3_update_vim..<SNR>3_tick..<SNR>3_spawn:
    line   22:
    E171: Missing :endif
    Error detected while processing function <SNR>3_update..<SNR>3_update_impl..<SNR>3_update_vim..<SNR>3_tick:
    line   28:
    E171: Missing :endif
    Error detected while processing function <SNR>3_update..<SNR>3_update_impl:
    line  106:
    E171: Missing :endif

**Comment by junegunn** [comment]
`:version`?

**Comment by orlp** [comment]
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Aug 25 2014 03:25:40)
    MS-Windows 64-bit GUI version with OLE support
    Included patches: 1-417
    Compiled by Haroogan <Haroogan@gmail.com>
    Huge version with GUI.  Features included (+) or not (-):
    +acl                +cmdline_compl      +diff               +file_in_path       +keymap             +modify_fname       +persistent_undo    +scrollbind         -tag_any_white      +visual             -xterm_save
    +arabic             +cmdline_hist       +digraphs           +find_in_path       +langmap            +mouse              -postscript         +signs              -tcl                +visualextra        +xpm_w32
    +autocmd            +cmdline_info       -directx            +float              +libcall            +mouseshape         +printer            +smartindent        -tgetent            +viminfo            
    +balloon_eval       +comments           -dnd                +folding            +linebreak          +multi_byte_ime/dyn +profile            -sniff              -termresponse       +vreplace           
    +browse             +conceal            -ebcdic             -footer             +lispindent         +multi_lang         +python/dyn         +startuptime        +textobjects        +wildignore         
    ++builtin_terms     +cryptv             +emacs_tags         +gettext/dyn        +listcmds           -mzscheme           +python3/dyn        +statusline         +title              +wildmenu           
    +byte_offset        +cscope             +eval               -hangul_input       +localmap           +netbeans_intg      +quickfix           -sun_workshop       +toolbar            +windows            
    +cindent            +cursorbind         +ex_extra           +iconv/dyn          +lua/dyn            +ole                +reltime            +syntax             +user_commands      +writebackup        
    +clientserver       +cursorshape        +extra_search       +insert_expand      +menu               +path_extra         +rightleft          +tag_binary         +vertsplit          -xfontset           
    +clipboard          +dialog_con_gui     +farsi              +jumplist           +mksession          +perl/dyn           +ruby/dyn           +tag_old_static     +virtualedit        -xim                
       system vimrc file: "$VIM\vimrc"
         user vimrc file: "$HOME\_vimrc"
     2nd user vimrc file: "$HOME\vimfiles\vimrc"
     3rd user vimrc file: "$VIM\_vimrc"
          user exrc file: "$HOME\_exrc"
      2nd user exrc file: "$VIM\_exrc"
      system gvimrc file: "$VIM\gvimrc"
        user gvimrc file: "$HOME\_gvimrc"
    2nd user gvimrc file: "$HOME\vimfiles\gvimrc"
    3rd user gvimrc file: "$VIM\_gvimrc"
        system menu file: "$VIMRUNTIME\menu.vim"
    Dependency: python27.dll, python34.dll, x64-msvcrt-ruby200.dll, lua52.dll, perl518.dll, libintl.dll, libiconv.dll, iconv.dll

**Comment by junegunn** [comment]
Thanks, also please post a 3-line vimrc that can reproduce the problem. (`plug#begin`, `Plug`, `plug#end`)

**Comment by orlp** [comment]
@junegunn 

    call plug#begin()
    Plug    'junegunn/vim-easy-align'
    call plug#end()

**Comment by junegunn** [comment]
Thanks I'll take a look when I get a chance. (I don't have easy access to a windows dev machine.)

**Comment by orlp** [comment]
I've done a binary search, and this is the first commit that breaks: https://github.com/junegunn/vim-plug/commit/9abd50facdf4318e15454e271b87c12ddedd21d6

**Comment by orlp** [comment]
The above commit of course only allows the symptoms to show - the real breakage occurs for me since this commit: https://github.com/junegunn/vim-plug/tree/59748b0680fc8be22c38bad7797037520596f57d .

**Comment by orlp** [comment]
Never mind about https://github.com/junegunn/vim-plug/commit/9abd50facdf4318e15454e271b87c12ddedd21d6 working - it only seems to work (it produces no output), but doesn't __actually__ update the repositories.

**Comment by starcraftman** [comment]
@orlp I just tested on my Windows 7 machine with Vim 32 & I've no problem with latest vim-plug. I'm using this build of VIM > http://tuxproject.de/projects/vim/complete-x86.7z . I only tested default vimL installer, not ruby.

 Also, looking at your stack trace, it seems the system command failed (or with symptom) that it couldn't write a tmp file. Seen this before, don't remember exact set of causes. Will have to Google.

```vim
E484: Can't open file C:\Users\Orson\AppData\Local\Temp\VIo5763.tmp
```

My guy says a local issue, but that's highly unscientific. I figure another Windows user would have complained if we'd been broken for 2 months. At least I hope there's more than 2 of us...

**Comment by starcraftman** [comment]
@orlp See this > http://vim.wikia.com/wiki/Fix_errors_that_relate_to_reading_or_creating_files_in_the_temp_or_tmp_environment_on_an_MS_Windows_PC

Knew I'd seen it before. First try changing the value of $TMP & $TEMP inside vim to some other folder you know you have write access & space. See if that fixes.

**Comment by orlp** [comment]
@starcraftman If I do `let $TMP='d:/tmp' I get the same error (albeit with a different path).

**Comment by starcraftman** [comment]
@orlp I see, the other thing they mention is shell escaping, our favorite friend. Do you have special chars they mention in the path you pass to plug#begin()? Any of these ` ^&|<>() `? I remember we fixed for spaces.

**Comment by orlp** [comment]
@starcraftman Yes, `C:\Program Files (x86)\dotfiles\vim\bundle`.

**Comment by starcraftman** [comment]
@orlp Can you try using `C:\Users\username\vim\plugged` out of curiosity?

**Comment by orlp** [comment]
@starcraftman Same issue.

**Comment by starcraftman** [comment]
@orlp Interesting, those were the obvious things I was thinking ruled out. Can you give me the output of the following ex commands.
```
echo &shell
echo &shellcmdflag
```

**Comment by orlp** [comment]
C:\Windows\system32\cmd.exe
    /c

**Comment by orlp** [comment]
Oh boy, I just figured it out. This is going to be a bit of a long story.

**Comment by orlp** [comment]
Scrap that - didn't figure it out - was accidently doing `:PlugInstall` rather than `:PlugUpdate` to test >.<

**Comment by starcraftman** [comment]
@orlp Just to confirm, you mentioned commit `59748b0` is the one that broke it. I looked at commit, I would like to confirm where the problem is. Can you apply this patch to function s:with_cd to your plug.vim . Does it work after?

```diff
function! s:with_cd(cmd, dir) 
-  return printf('cd%s "%s" && %s', s:is_win ? ' /d' : '', escape(a:dir, '"'), a:cmd)
+  return (s:is_win ? 'cd /d ' : 'cd ').s:esc(a:dir).' && '.a:cmd
```

**Comment by orlp** [comment]
@starcraftman Yes and no. The error I described above in this issue is gone, but this issue comes back: https://github.com/junegunn/vim-plug/issues/154 .

**Comment by starcraftman** [comment]
@orlp So it is the Windows escaping code. Le sigh. Not helpful that I can't reproduce this new problem.

**Comment by orlp** [comment]
@starcraftman I see that you guys do a lot of manual escaping - what's wrong with `shellescape()`?

**Comment by junegunn** [comment]
@orlp See #13

**Comment by orlp** [comment]
@junegunn Isn't that avoided by the classic `set cpo`?

**Comment by junegunn** [comment]
@orlp Not sure. Last time I checked, shellescape does more than what we needed and caused more harm than good. If you can simplify the code using shellescape, please send us a patch.

**Comment by junegunn** [comment]
@orlp 

> In reality both parallel and serial updates seem to have been broken since 59748b0 .

I'm a bit confused about this. 59748b0 was the fix to #154 that you reported and you confirmed that all is fixed with the patch. And now you say it is what currently causing the error. What happened to your environment?

**Comment by orlp** [comment]
@junegunn  Good question, I don't know =/

**Comment by starcraftman** [comment]
@orlp Well that does hinder fixing things, not knowing what has changed. Hard to debug this without local reproduction.

One to try, can you make a new Windows user on this system to isolate local issues? If so, then use that one & download the tuxbuild and setup vim-plug with following vimrc. Then try again.
```vim
call plug#begin()
Plug 'junegunn/vim-easy-align'
call plug#end()
```

If it still fails, only other thing I see is one of us RDPing onto your box to debug by hand.

@junegunn I did some extra testing on special chars in paths by passing strings to plug#begin(). The only chars we don't handle well on Windows are `<>|`. If you put those into a path, the leaf of the path will be ignored. That is, when I did something like `plug#begin('$HOME/vim/test<')` would install to '$HOME/vim' and ignore test folder.

**Comment by orlp** [comment]
@starcraftman I seem to have found a fix for the issues I'm facing.

If I change this line (removing parentheses around the command):

    \ printf('(git fetch %s %s 2>&1 && git checkout -q %s 2>&1 && git merge --ff-only %s 2>&1 && git submodule update --init --recursive 2>&1)',

to 

    \ printf('git fetch %s %s 2>&1 && git checkout -q %s 2>&1 && git merge --ff-only %s 2>&1 && git submodule update --init --recursive 2>&1',

the symptoms seem to disappear. This is using the latest version of vim-plug. I don't know what kind of effect that change will have on other systems.

However, even if I set `g:plug_threads` to 16 it won't do parallel updates.

**Comment by orlp** [comment]
The plot thickens...

If I do this above fix it works for my full vimrc, but breaks when doing a simple three-line vimrc.

**Comment by junegunn** [comment]
@orlp Correct me if I'm wrong but as far as I know, without the parens, Vim receives only the output of the last command (`git submodule update`) on Windows. And I do not wish to change anything until you can answer [my previous question](https://github.com/junegunn/vim-plug/issues/229#issuecomment-102215798): what has changed? Can we be sure it'll not change again? I'd like you to try what @starcraftman suggested above.

> Although I still don't understand why parallel updates don't work - was it disabled?

No, the default number of threads on windows was [changed to 1](9abd50facdf4318e15454e271b87c12ddedd21d6) as the Ruby installer on Windows is really unstable, but if you have set `g:plug_threads` or run `PlugInstall SOME_NUMBER`, it should work. If it doesn't, it's a bug.

@starcraftman 
> @junegunn I did some extra testing on special chars in paths by passing strings to plug#begin(). The only chars we don't handle well on Windows are <>|. If you put those into a path, the leaf of the path will be ignored. That is, when I did something like plug#begin('$HOME/vim/test<') would install to '$HOME/vim' and ignore test folder. Not hugely likely somebody uses those though.

Interesting. Thanks for the heads-up, I'll take note. As you said, I don't think anyone would try to use those chars in the path. We'll see :)

**Comment by orlp** [comment]
@junegunn I already answered your question, albeit with a very unsatisfying answer: I don't know what changed.

I'll try creating a custom user, but if that doesn't work I'm O.K. with one of you guys RPD'ing into my machine if you are willing to spend the time.

**Comment by orlp** [comment]
You are right though that without `()` vim only returns the output of the last command - I wasn't aware. So my fix is not right.

**Comment by orlp** [comment]
I figured out what in my environment that triggers the bug. There is a registry key in Windows called `KCU\Software\Microsoft\Command Processor\Autorun`. If you set this key to a batch file, `cmd.exe` will always run this batch file before executing its command.

I use this batch file to set up my `cmd.exe` environment on Windows, similar to `.bashrc`. I have this at the top of my `setup_environment.bat`:

    REM don't bother setting up a nice environment if we're not interactive
    echo %CMDCMDLINE% | C:\Windows\System32\find.exe /i "/c" >nul
    if not errorlevel 1 goto exit

For some reason this check fails, and then something messes with the command line enough to error out Vim.

**Comment by starcraftman** [comment]
@orlp Interesting. Never used that. I don't like that such a thing is tucked away in the registry, then again I dislike the registry entirely. I looked at cmd.exe's flags and they say /D disables the autorun files from the registry so I guess that's our patch for this. Can you open up vim & run this command:
let &shellcmdflag='/D /c'. Then try updating. It should work. 

I'll make a patch in a bit, I'm hungry.

**Comment by orlp** [comment]
@starcraftman Well, I'm not certain if that's what you really want. Some people might set, for example, `git` using the autorun system. Not saying it's too sane, but it might happen.

I think the problem lies with my batch file. I'll have to fix my end. Which I'm not going to do because I'll simply be not be using this system anymore, but only setting up the environment in my custom console. I no longer use cmd.exe directly.

---

However, one issue still isn't fixed. Nothing runs in parallel, despite having set `let g:plug_threads = 16` before calling `plug#begin`, and having both `python/dyn` and `python3/dyn`  (and tested, `:py print(1)` and `:py3 print(1)` work).

**Comment by starcraftman** [comment]
@orlp Re parallel, the python installer only works on POSIX, sorry :/. Ruby one works somewhat but no screen updates during. So you should check if `ruby puts 'hello'` works. Both installers unfortunately suffer from #81 which is why you don't get any updates during the command. Due to 81 & lack of a good portable/builtin alternative to pgroup kill on Windows I just didn't enable it.

You may have a point regarding /D, but I'd say its REALLY bad practice if anyone relied on an autorun instead of %PATH% to find their git. 

The possibilities I see are:
1) Document on wiki faq & don't do anything.
2) Use `/D` shell flag & if we break someone (somewhat unlikely) we deal with them.
3) Query at startup if on windows if these reg keys are set (using something like:  `REG QUERY HKCU\Software\Microsoft\Command Processor /v Autorun`

What do you think @junegunn ?

**Comment by junegunn** [comment]
@starcraftman Hmm. Option 3: definitely not. It feels like throwing in a bunch of code that is too system-specific and will not be well tested and maintained, especially considering that it's not something we can test in our automated test scenario. Then option 1 vs 2. I'm okay with 1, since @orlp, who's having the problem, agrees that the problem is in the batch file rather than in vim-plug. But we could still consider option 2, if it costs only a few lines of code (temporarily setting the option and restoring it), and if it's generally considered to be a better practice on Windows. Is it? I can't tell since I know nothing about Windows environment. Not being able to test it bother me though.

**Comment by starcraftman** [comment]
@junegunn I thought about it & I think its best we just document. It probably isn't many win users using this & so no point added more code we don't routinely test. I've added a section on the [faq](https://github.com/junegunn/vim-plug/wiki/faq) that documents what happens & queries to see if affected. So we can close this for now.

**Comment by junegunn** [comment]
@starcraftman Great, thank you very much!

**Comment by orlp** [comment]
@junegunn @starcraftman 

Sorry for bothering you with what was an issue on my side. As a defense, it was really obscure :)

**Comment by starcraftman** [comment]
@orlp Its ok, I learned about AutoRun at least. I will be wary of that in future. I have this sinking feeling like that's been used as an attack vector when combined with phishing. People click anything.
